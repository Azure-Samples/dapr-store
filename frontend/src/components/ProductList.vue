<template>
  <h1>PRODUCT LIST {{ viewType }}</h1>
</template>

<script>
export default {
  name: 'ProductList',

  props: {
    viewType: {
      type: String,
      required: true
    }
  }
}
</script>

<style scoped>

</style>