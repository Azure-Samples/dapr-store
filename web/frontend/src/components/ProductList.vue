<!--
// ----------------------------------------------------------------------------
// Copyright (c) Ben Coleman, 2020
// Licensed under the MIT License.
//
// Dapr Store frontend - Reusable component that lists products
// ----------------------------------------------------------------------------
-->

<template>
  <div>
    <div v-if="!products" class="text-center">
      <b-spinner variant="success" style="width: 5rem; height: 5rem;" />
    </div>

    <b-card v-for="product in products" :key="product.id">
      <b-row>
        <b-col>
          <b-link :to="`/product/`+product.id">
            <b-card-title>
              {{ product.name }}
            </b-card-title>
          </b-link>
          <b-card-text>
            {{ product.description }}
            <br><br>
            <h4>Â£{{ product.cost }}</h4>
          </b-card-text>

          <b-button :disabled="!isLoggedIn()" href="#" variant="primary" class="d-none d-md-inline" @click="addToCart(product)">
            <fa icon="shopping-cart" />
            &nbsp; Add to Cart
          </b-button>
        </b-col>

        <b-col class="flex-grow-0 d-none d-md-block">
          <div class="product-img">
            <span v-if="product.onOffer" class="onsale">On Sale</span>
            <b-link :to="`/product/`+product.id">
              <img :src="product.image">
            </b-link>
          </div>
        </b-col>
      </b-row>
    </b-card>
  </div>
</template>

<script>
import api from '../services/api'
import auth from '../services/auth'

export default {
  name: 'ProductList',


  props: {
    products: {
      type: Array,
      default: () => []
    }
  },

  methods: {
    async addToCart(product) {
      try {
        if (!auth.user()) { return }

        await api.cartAddAmount(auth.user().userName, product.id, +1)
        this.showToast('Added to your cart!', 'success', product)
      } catch (err) {
        this.showToast('Error adding to cart ðŸ˜« '+err.toString(), 'danger', product)
      }
    },

    isLoggedIn() {
      if (auth.user()) { return true }
      return false
    },

    showToast(msg, variant, product) {
      this.$bvToast.toast(`${product.name}`, {
        title: msg,
        variant: variant,
        autoHideDelay: 3000,
        appendToast: true,
        toaster: 'b-toaster-top-center',
        solid: true
      })
    }
  }
}
</script>

<style scoped>
.card {
  margin: 1rem;
}
.card-title {
  font-size: 2rem;
}
.product-img {
  float: right;
}
.product-img img {
  width: 12rem;
  border-radius: 0.5rem;
}
.card-text {
  font-size: 130%
}
.onsale {
  display: inline-block;
  position: absolute;
  bottom: 0;
  width: 12rem;
  height: 1.5rem;
  font-size: 1rem;
  color: white;
  background-color: rgb(185, 17, 17);
  text-align: center;
  border-bottom-left-radius: 0.5rem;
  border-bottom-right-radius: 0.5rem;
}
</style>