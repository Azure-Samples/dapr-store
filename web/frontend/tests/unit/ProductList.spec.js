import { createLocalVue, mount } from '@vue/test-utils'
import flushPromises from 'flush-promises'
import BootstrapVue from 'bootstrap-vue'
const localVue = createLocalVue()
localVue.use(BootstrapVue)

import VueRouter from 'vue-router'
localVue.use(VueRouter)
const router = new VueRouter()

import ProductList from '@/components/ProductList.vue'

jest.mock('@/services/api')

// Load mock data
let mockJson = require('fs').readFileSync(__dirname+'/../../../../testing/mock-data/products.json')
let mockProducts = JSON.parse(mockJson)

describe('ProductList.vue', () => {
  it('renders product list', async () => {
    const wrapper = mount(ProductList, {
      localVue,
      router,
      propsData: {
        products: mockProducts
      }
    })

    await flushPromises()
    expect(wrapper).toMatchSnapshot()
  })
})