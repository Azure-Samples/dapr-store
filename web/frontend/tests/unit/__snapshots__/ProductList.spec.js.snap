// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList.vue renders product list 1`] = `
<div>
  <!---->
  <div class="card">
    <!---->
    <!---->
    <div class="card-body">
      <!---->
      <!---->
      <div class="row">
        <div class="col"><a href="#/product/1" class="" target="_self">
            <h4 class="card-title">
              Top&nbsp;Hat&nbsp;(6″)
            </h4>
          </a>
          <p class="card-text">
          <h4>Made from 100% Wool and nice</h4>

          £39.95
          </p> <a role="button" aria-disabled="true" tabindex="-1" target="_self" href="#" class="btn d-none d-md-inline disabled btn-primary disabled"><i icon="shopping-cart"></i>
            &nbsp; Add to Cart
          </a>
        </div>
        <div class="flex-grow-0 d-none d-md-block col">
          <div class="product-img"><a href="#/product/1" class="" target="_self"><img src="/img/catalog/1.jpg"></a></div>
        </div>
      </div>
    </div>
    <!---->
    <!---->
  </div>
  <div class="card">
    <!---->
    <!---->
    <div class="card-body">
      <!---->
      <!---->
      <div class="row">
        <div class="col"><a href="#/product/2" class="" target="_self">
            <h4 class="card-title">
              Black &amp; Gold Tie Set
            </h4>
          </a>
          <p class="card-text">
          <h4>100% hand made silk includes tie, pocket square and cufflinks</h4>

          £18
          </p> <a role="button" aria-disabled="true" tabindex="-1" target="_self" href="#" class="btn d-none d-md-inline disabled btn-primary disabled"><i icon="shopping-cart"></i>
            &nbsp; Add to Cart
          </a>
        </div>
        <div class="flex-grow-0 d-none d-md-block col">
          <div class="product-img"><a href="#/product/2" class="" target="_self"><img src="/img/catalog/2.jpg"></a></div>
        </div>
      </div>
    </div>
    <!---->
    <!---->
  </div>
  <div class="card">
    <!---->
    <!---->
    <div class="card-body">
      <!---->
      <!---->
      <div class="row">
        <div class="col"><a href="#/product/3" class="" target="_self">
            <h4 class="card-title">
              Paisley Cravat Ascot Tie
            </h4>
          </a>
          <p class="card-text">
          <h4>Black &amp; Red Paisley Cravat</h4>

          £11.2
          </p> <a role="button" aria-disabled="true" tabindex="-1" target="_self" href="#" class="btn d-none d-md-inline disabled btn-primary disabled"><i icon="shopping-cart"></i>
            &nbsp; Add to Cart
          </a>
        </div>
        <div class="flex-grow-0 d-none d-md-block col">
          <div class="product-img"><a href="#/product/3" class="" target="_self"><img src="/img/catalog/6.jpg"></a></div>
        </div>
      </div>
    </div>
    <!---->
    <!---->
  </div>
</div>
`;
